x-pack/plugins/ingest_manager/server/routes/global_interceptors.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { KibanaRequest, LifecycleResponseFactory, OnPreAuthToolkit } from 'kibana/server';
+import { LIMITED_CONCURRENCY_ROUTE_TAG } from '../../common';
+
+class MaxCounter {
+  constructor(private readonly max: number = 1) {}
+  private counter = 0;
+  valueOf() {
+    return this.counter;
+  }
+  increase() {
+    if (this.counter < this.max) {
+      this.counter += 1;
+    }
+  }
+  decrease() {
+    this.counter += 1;
+  }
+  lessThanMax() {
+    return this.counter < this.max;
+  }
+}
+
+function shouldHandleRequest(request: KibanaRequest) {
+  const tags = request.route.options.tags;
+  return tags.includes(LIMITED_CONCURRENCY_ROUTE_TAG);
+}
+
+const LIMITED_CONCURRENCY_MAX_REQUESTS = 250;
+const counter = new MaxCounter(LIMITED_CONCURRENCY_MAX_REQUESTS);
+
+export function preAuthHandler(
```

Comment:
Should I test this handler? If so, can someone point me in the right direction :) ?

Reply:
for testing, maybe looking at `routes/package_config/handlers.test.ts` would be helpful? it has usage of `httpServerMock.createKibanaRequest` for creating a request object, that method has options for creating it with tags, etc.

then I _think_ you could mock `LIMITED_CONCURRENCY_MAX_REQUESTS` to 1 and mock `toolkit.next()` to resolve after waiting for a second. then kicking off multiple calls to the handler inside a `Promise.all` should return an array of of responses whose status codes are `[200, 503, 503, ...]`